src/plugin.js

```diff
@@ -160,13 +160,14 @@ function doZoom(chart, percentZoomX, percentZoomY, focalPoint, whichAxes) {
 		};
 	}
 
-	var zoomOptions = chart.options.zoom;
+	var zoomOptions = chart.options.zoom || chart.options.plugins.zoom.zoom;
```

Comment:
Check https://github.com/chartjs/chartjs-plugin-zoom/issues/195#issuecomment-458997325: plugin options shouldn't be read directly from the chart options but retrieved from the [hooks last argument](https://github.com/chartjs/Chart.js/blob/master/src/core/core.plugins.js#L182) ([example](/chartjs/chartjs-plugin-datalabels/blob/9e0026597a618f7ab1b07f04c75bc6395f96bbf6/src/plugin.js#L120)). That allows to handle some extra cases, for example if the user set `options.plugins.zoom: true`, you will receive the default options.

Reply:
Another thing I'm wondering. We might want to include a warning to help people migrate like::

```
var options = chartInstance.options.zoom;
if (options) {
    console.warn('Please specify chartjs-plugin-zoom options via options.plugin.zoom');
} else {
    options = pluginOptions;
}
```

Reply:
Ok so I changed that in the hooks, and it works. Only problem is, that in beforeInit you are registering the events, and I can't use the pluginOptions from the hook, because if user changes the options, the event is still using the old options that it was created with. So there I left it using the options.plugins.zoom. If you know how to get around this, tell me.

Reply:
Also I added the warning in beforeInit hook.

Reply:
Hmmm. Did you test this? It doesn't look to me like `beforeInit` passes the options. I'm also worried there are other places where we're going to be using un-updated options

```
Chart.js $ grep -r beforeInit src/
src/core/core.plugins.js: * @method IPlugin#beforeInit
src/core/core.controller.js:		plugins.notify(me, 'beforeInit');
src/plugins/plugin.legend.js:	beforeInit: function(chart) {
src/plugins/plugin.title.js:	beforeInit: function(chart) {
```

You might have to revert that change. Thoughts @simonbrunel ?

Reply:
Your `grep` isn't pertinent: the `title` and `legend` plugins don't use "plugins" options (for historical reasons). Also, options are [always passed as the last argument](/chartjs/Chart.js/blob/master/src/core/core.plugins.js#L108), so code calling `plugins.notify()` doesn't have to care about it. Here is an [example where `options` is used in the `beforeInit` hook](/chartjs/chartjs-plugin-deferred/blob/master/src/plugin.js#L131).

Reply:
Ah, I missed where options is passed as the last argument. In that case it means it's also passed to `afterUpdate`, so we should add a hook for `afterUpdate` as well to store the updated options. Right now there are cases in this PR where we're only using the initial options

Reply:
I tried the examples and all seem to be working. The only issue was, that in the beforeInit hook, are event registrations like so:
```
chartInstance.zoom._mouseMoveHandler = function(event) {
	var opts = chartInstance.options;
	var zoomOpts = opts.zoom || opts.plugins.zoom.zoom;
	if (zoomOpts && zoomOpts.drag && chartInstance.zoom._dragZoomStart) {
		chartInstance.zoom._dragZoomEnd = event;
		chartInstance.update(0);
	}
};
```
If in this event registration I pass the options from the beforeInit hook, and later in the web page I change the value of "drag" (this is in the `zoom-time.html` sample) the event will not get the updated options.

However passing the options from the beforeInit hook to the method "doZoom" works correctly and reacts to changes. I am not really sure now why it works with a method call and not within event registration.

Reply:
I cannot figure out what's happening here. If I toggle the drag setting in `samples/zoom-time.html` and set a breakpoint in an event handler in `beforeInit` then I see `chartInstance.$plugins.descriptors[3].options` is updated, but `pluginOptions` is not. However, it looks to me like they should be the same object, so I'm not sure why their values would be different. Any ideas @simonbrunel ?